### PR TITLE
[Cognitive Services] Support ApiProperties which enable creating QnAMaker

### PR DESCRIPTION
**Pull request by yangyuan**
Add a new parameter --api-properties for create command.
--api-properties could be either key=value key=value pairs or JSON string.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yangyuan** [comment]
I will take a look at the failed checks.

**Comment by yangyuan** [comment]
Hi @tjprescott, can you take a look at this :)

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py [hunk 1]
```diff
@@ -3,17 +3,53 @@
...
+def _is_suspected_json(string):
+    """ If the string looks like a JSON """
+    if string.startswith('{') or string.startswith('\'{') or string.startswith('\"{'):
+        return True
+    if string.startswith('[') or string.startswith('\'[') or string.startswith('\"['):
+        return True
+    if re.match(r"^['\"\s]*{.+}|\[.+\]['\"\s]*$", string):
+        return True
+
+    return False
```
Comment (tjprescott) [inline src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: There is a function in core called "shell_safe_json_parse" that does a lot of this plus some other stuff. See if that will work for your purpose.  You basically attempt the JSON parse and if it fails, treat it as a string.

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py [hunk 2]
```diff
@@ -3,17 +3,53 @@
...
+
+
+api_properties_type = CLIArgumentType(
+    type=validate_api_properties,
+    help="Api properties in JSON format or a=b;c=d format. Some cognitive services (i.e. QnA Maker) "
```
Comment (tjprescott) [inline src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: The command-line format should be like tags or metadata: i.e `a=b c=d`. Space-separated key values pairs, not comma or semi-colon separated. You would basically use a validator that looks like this (pulled from Network):

```Python
def validate_metadata(namespace):
    if namespace.metadata:
        namespace.metadata = dict(x.split('=', 1) for x in namespace.metadata)
```

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 1 above)
Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: Not exactly. 

Attempt the JSON parse
-> success, treat it as JSON
-> failed, call this method, then treat it as string:
     -> if this method say yes: give a warning.

Reason:
CLI can accept following formats:
```
{\"a\":\"b\"}
"{\"a\":\"b\"}"
```

But very likely the customer will try below formats:
```
{"a":"b"}
'{"a":"b"}'
'{\"a\":\"b\"}'
```

And I want to give warning is customer did that.

But `shell_safe_json_parse` should be useful to replace  json.loads, but I will have to catch CLIError

Comment (tjprescott) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: Generally catching the CLIError is where you will proceed assuming it isn't JSON.

**Comment by yangyuan** [comment]
Hi @tjprescott, I've update the JSON parsing part to use shell_safe_json_parse. Also let api-properties to support space separated kay=value pairs

```
if is list:
-> must be key=value pairs
attempt the JSON parse:
-> success, treat it as JSON
-> failed, call _is_suspected_json method, then treat it as a single key=value pair:
     -> if this method say yes: give a warning.
```

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 1 above)
Comment (troydai) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: This doesn't make it similar. We shouldn't validate this in the CLI. Different terminals have different approaches of escaping characters and the attempt to consolidate them in the CLI will complicate the text processing. We should simply threat the input as string and let `json` module to parse it and return whatever error message it signals.

**Comment by tjprescott** [comment]
@troydai I have no idea why GitHub will not let me reply to your comment. Annoying. Anyways, I agree that this code shouldn't be doing a bunch of custom parsing. however, it should rely on shell_safe_json_parse and not just json directly.

**Comment by yangyuan** [comment]
Hi @troydai:
I don't understand your point. That function is just for determining if we wants to giving warnings. It does not participate in parsing JSON.

**Comment by troydai** [comment]
ok, i got it.

**Pull request merged**